File: 701-insert-into-a-binary-search-tree/701-insert-into-a-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode insertIntoBST(TreeNode root, int val) {
        if(root == null){
            return new TreeNode(val);
        }
        
        TreeNode left; 
        TreeNode right; 
        
        if(val > root.val){
            
            right = insertIntoBST(root.right, val);
            left = root.left;
        }else{
            left = insertIntoBST(root.left, val);
            right = root.right;
        }
        root.left = left;
        root.right = right;
        
        return root;
    }
}

File: 701-insert-into-a-binary-search-tree/701-insert-into-a-binary-search-tree.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @param {Integer} val
# @return {TreeNode}
def insert_into_bst(root, val)
    #first case
    if(root == nil)
        return TreeNode.new(val, nil, nil)
    elsif(val < root.val)
        root.left = insert_into_bst(root.left, val)
    else
        root.right = insert_into_bst(root.right, val) 
    end
    root
end